### PR TITLE
📖 clarify providerID docs

### PR DESCRIPTION
**Pull request by alexeldeib**
**What this PR does / why we need it**:
improve providerID docs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

https://github.com/kubernetes-sigs/cluster-api/issues/3007

### CODE REVIEW

File: docs/book/src/developer/architecture/controllers/machine.md [hunk 1]
```diff
@@ -14,6 +14,17 @@ The Machine controller's main responsibilities are:
...
 * Keeping the Machine's Status object up to date with the InfrastructureMachine's Status object.
+* Finding Kubernetes nodes matching the expected providerID in the workload cluster.
+
+After the machine controller sets the OwnerReferences on the associated objects, it waits for the bootstrap
+and infrastructure objects referenced by the machine to have the `Status.Ready` field set to `true`. When 
+the infrastructure object is ready, the machine controller will attempt to read its `Spec.ProviderID` and
+copy it into `Machine.Spec.ProviderID`.
+
+The machine controller uses the kubeconfig for the new workload cluster to watch new nodes coming up.
+When a node appears with `Node.Spec.ProviderID` matching `Machine.Spec.ProviderID`, the machine controller
```
Comment (JoelSpeed) [inline docs/book/src/developer/architecture/controllers/machine.md:25]: This matching isn't currently true in most of the cloud providers, are you aware of this? We should make sure there are issues for tracking this in the various provider repos to make sure they get updated

Comment (enxebre) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: For ref https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1693

Comment (alexeldeib) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: I’m not following the connection between the linked issue and this comment? I wrote this blurb based on my read of the generic machine controller. It’s not based on any of the infra provider implementations. What am I missing?

Comment (alexeldeib) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: How does this logic work for CAPA if this text is wrong? https://github.com/kubernetes-sigs/cluster-api/blob/df9fcf9442ee353d2b388f8d79361a75d21ec487/controllers/machine_controller_noderef.go#L49-L83

What *is* correct here then?

Comment (JoelSpeed) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: Perhaps I'm being overly pedantic when I read "matching". My comment was more of an FYI than there's anything wrong with this, in fact, I totally agree with the changes you've made to this description, this is what I thought was happening and what I thought should be happening too.

However, in the real case, when you spin up a cluster-api cluster on AWS (or others, just using AWS as an example), the `machine.Spec.ProviderID` will looking like `aws:////<instance-id>`, whereas when the node joins the cluster, the Kubernetes cloud provider will set the `node.Spec.ProviderID` to `aws:///<availability-zone>/<instance-id>`. So technically, they don't match, as in, they aren't identical, but they can be matched. We have been having a discussion about fixing this in CAPA, but need to raise the point in other providers as well as this not matching breaks the core Cluster Autoscaler's assumptions meaning autoscaling with CAPI doesn't work currently

File: docs/book/src/developer/architecture/controllers/machine.md [hunk 2]
```diff
@@ -62,14 +73,19 @@ status:
...
 ### Infrastructure provider
 
-The InfrastructureMachine object **must** have a `status` object.
+The InfrastructureMachine object **must** have both `spec` and `status` objects.
+
+#### Required `spec` fields
+
+The `spec` object **must** at least one field defined:
+
+* `providerID` - a cloud provider ID identifying the machine. This is often set by the cloud-provider-controller.
```
Comment (vincepri) [inline docs/book/src/developer/architecture/controllers/machine.md:82]: What do you mean by `cloud-provider-controller` in this context?

File: docs/book/src/developer/architecture/controllers/machine.md (hunk 1 above)
Comment (alexeldeib) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: Thanks for the info, that makes sense. I’m still a little confused how CAPA works around this today, does that not break machine controller as well as CA?

Comment (JoelSpeed) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: I think they have a regex that checks the format and then the method discards any additional information other than the `<instance-id>` when doing the check

https://github.com/kubernetes-sigs/cluster-api/blob/df9fcf9442ee353d2b388f8d79361a75d21ec487/controllers/noderefutil/providerid.go#L38-L73

Comment (vincepri) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: Maybe we should reword or define how we match? We now only look at the last part, the `ProviderSpecificNodeID`, per https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3741

Comment (JoelSpeed) [reply docs/book/src/developer/architecture/controllers/machine.md:25]: Rewording how it's matched would make it accurate for now, but I think the other other providers should implement exact matching in the future so we could keep it as is and open issues to track changing the providers? This probably is worth a discussion on the community call

File: docs/book/src/developer/architecture/controllers/machine.md (hunk 2 above)
Comment (alexeldeib) [reply docs/book/src/developer/architecture/controllers/machine.md:82]: typo, infra provider controller

Comment (vincepri) [reply docs/book/src/developer/architecture/controllers/machine.md:82]: I'd just say `infrastructure providers`, like we do elsewhere

Comment (alexeldeib) [reply docs/book/src/developer/architecture/controllers/machine.md]: actually just deleting this, it's clarified above.

**Comment by vincepri** [comment]
LGTM, @JoelSpeed ?

**Comment by vincepri** [comment]
@alexeldeib can you squash commits?

**Comment by JoelSpeed** [comment]
I'm happy with the content after squash 👍

**Comment by fabriziopandini** [comment]
as per comments above
/lgtm

**Pull request merged**